### PR TITLE
R4R: StateDB Integration/EVM Module Start

### LINKED ISSUE #486: R&D: StateDB/EVM Integration with SDK modules
Attempting to integrate and use Geth as an extended library within Ethermint has surfaced some challenges, namely in regards to bridging the gap between the StateDB/EVM layer in Geth and the SDK's modules when using accounts and coins. We'd like to utilize a single coin to facilitate staking and governance functionality and to also be used within dApps.

This goal will ultimately result in having to maintain a forked version of Geth which has it's own StateDB implementation or utilize another client all together (i.e. Turbo-Geth).

ref: #482

/cc @AlexeyAkhunov

### PR DESCRIPTION
**Pull request by alexanderbez**
- Forked Geth to turn `state.StateDB`  and `state.stateObject` into interfaces
- Implement necessary types and interfaces within Ethermint in order to fully integrate SDK components (stores and account mapper) with the StateDB implementation

I also updated to the latest SDK which caused additional changes. Major things to review are under `x/evm`.

Ultimately I do not think it's wise to maintain a fork of Geth, but this will be a good reference point to see what we need to tweak in Turbo-Geth to use that as a client.

closes: #486 

/cc @jackzampolin @jaekwon @AlexeyAkhunov 

----

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ethermint/blob/master/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added entries in `PENDING.md` with issue # 
- [x] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]"

### CODE REVIEW

**Comment by alexanderbez** [comment]
Updates:

- ~~We'll need the SDK to expose/provide the Merkle root (IAVL+ tree version/commitID hash) from the context in order to properly set tx receipts. I realized that the `lastCommitID` is not actually updated until we finally commit (BaseApp commit not StateDB commit). So the semantics from Ethereum's trie and this value don't exactly translate.~~
  - Update, realized we actually don't need this anymore. Ignore!


- I got a base implementation running and importing blocks but getting halted at block 46217 due to `insufficient balance to pay for gas` for this [TX](ethermine.org/miners/eA54d11A2aC4bAC8dB83DdA13AA0d7469C755406/). Looks like this is the first transaction of this account, so maybe I'm not properly creating new accounts?

In any case, these are the next two objects to tackle. Afterwards, it creating the EVM/StateDB module, more unit tests, wrapping things up and then finally hooking up TM. I recommend RPC work to hand in hand with this (@jackzampolin).

**Comment by jackzampolin** [comment]
Sounds good @alexanderbez, I'll get the ball rolling there. Thanks for the update!

**Comment by alexanderbez** [comment]
Figured out the issue was that I was not accumulating block rewards. I think I should have something ready soon to merge 👍

File: x/evm/types/dump.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package types
+
+import (
+	ethstate "github.com/ethereum/go-ethereum/core/state"
+)
+
+// RawDump returns a raw state dump.
+//
+// TODO: Implement if we need it, especially for the RPC API.
```
Comment (jackzampolin) [inline x/evm/types/dump.go:9]: This is a nice testing utility too...

**Review by jackzampolin — APPROVED** [review]

**Comment by jackzampolin** [comment]
Not a ton of comments here. I've pulled this down and run it against the `data/blockchain` file successfully.

**Pull request merged**